src/Components/WebAssembly/WebAssembly.Authentication/src/Interop/AuthenticationService.ts

```diff
@@ -52,28 +52,71 @@ export interface AuthenticationResult {
     message?: string;
 }
 
+export interface AuthenticationContext {
+    state?: unknown;
+    interactiveRequest: InteractiveAuthenticationRequest;
+}
+
+export interface InteractiveAuthenticationRequest {
+    scopes?: string[];
+    additionalRequestParameters?: { [key: string]: any };
+};
+
 export interface AuthorizeService {
     getUser(): Promise<unknown>;
     getAccessToken(request?: AccessTokenRequestOptions): Promise<AccessTokenResult>;
-    signIn(state: unknown): Promise<AuthenticationResult>;
+    signIn(context: AuthenticationContext): Promise<AuthenticationResult>;
     completeSignIn(state: unknown): Promise<AuthenticationResult>;
-    signOut(state: unknown): Promise<AuthenticationResult>;
+    signOut(context: AuthenticationContext): Promise<AuthenticationResult>;
     completeSignOut(url: string): Promise<AuthenticationResult>;
 }
 
+export class ManagedLogger {
+    private debug: boolean;
+    private trace: boolean;
+    public constructor(private objectRef: any) {
+        this.debug = objectRef.invokeMethod('IsEnabled', LogLevel.Debug);
+        this.trace = objectRef.invokeMethod('IsEnabled', LogLevel.Trace);
+    }
+
+    log(level: LogLevel, message: string): void {
+        if ((level == LogLevel.Trace && this.trace) ||
+            (level == LogLevel.Debug && this.debug)) {
+            return this.objectRef.invokeMethod('Log', level, message);
```

Comment:
TBH We're going to a lot of trouble to redirect the log output from TypeScript code to .NET code which then calls back into JS code (since this is wasm only). Wouldn't it be good enough just to log directly to the JS console from the original TypeScript code?

I know it's theoretically possible people may have set up something custom in the .NET logging config but that would be pretty unusual to do. Is that what motivates going through all the indirection?

At first glance I was a bit worried to see that we're allowing JS code to tell .NET to log a completely arbitrary message with an arbitrary severity. I quickly realised this is wasm only so not a security issue in itself, however we will have to remember not to extend the use of this pattern or the infrastructure you're adding here to Server in the future. It would be a bit safer if the infrastructure didn't exist in the first place.

Reply:
I am fine with doing just that, I somewhat wanted to keep the messages consistent, hence why I opted for this approach. I will simplify this to just pass in from .NET the enabled log levels (to keep things consistent) and log directly through the console in that case.